corehq/apps/users/management/commands/migrate_user_history_to_new_structure.py

```diff
@@ -0,0 +1,734 @@
+        assert change_messages, f"Change message not created for message {user_history.message}"
+        user_history.change_messages = change_messages
+
+    if save:
+        user_history.save()
+
+        migrated = migrate_user_history_to_log_entry(user_history)
+        assert migrated, f"Could not create log entry for User History record {user_history.pk}"
+
+
+def create_change_messages(user_history, skip_assertions):
+    """
+    A message is a ". " join of different messages.
+    So it should be possible to split by ". " and identify individual messages.
+    There are two messages that have ". " in it
+    - one for verified two factor disable
+        'Two factor disabled. Verified by: {verified_by}, verification mode: "{verification_mode}"'
+    - one for status update
+        'User {action}. Reason: "{reason}"'
+    They can be identified and handled accordingly
+    """
+
+    # split by . but not by something that starts with st like "St. " which was found in location names
+    # "St. " or "st. " was the only case found during the dry run
+    # if any message gets broken down poorly because of . in their messages,
+    # conversion won't work for such messages and they will be flagged
+    # all individual messages should have checks to ensure they are complete
+    messages = re.split(r"(?<!St)\. ", user_history.message, flags=re.IGNORECASE)
+    messages_not_converted = []
+    change_messages = {}
+
+    # flag for two factor disable message
+    disable_for_days = None
+
+    # flag for status update messages
+    status_update_active = None
+    status_update_reason = None
+
+    for message in messages:
+        if "Program: None" == message:
+            _check_for_double_entry(PROGRAM_FIELD, message, change_messages)
+
+            change_messages.update(
+                UserChangeMessage.program_change(None)
+            )
+
+        elif PROGRAM_REGEX.match(message):
+            _check_for_double_entry(PROGRAM_FIELD, message, change_messages)
+
+            program_name, program_id = PROGRAM_REGEX.match(message).groups()
+
+            assert program_name, f"Program name missing in message {message}"
+            assert program_id, f"Program id missing in message {message}"
+            program = Program.get(program_id)
+            assert program, f"Could not find program with id {program_id}"
+
+            if not skip_assertions:
+                # assert for complete message
+                # compare name
+                assert program.name == program_name, "Looks like program name changed. Found program id " \
+                                                     f"{program_id} with name {program.name} but name in log " \
+                                                     f"is {program_name}"
+
+            change_messages.update(UserChangeMessage.program_change(program))
+
+            # fallback in case program has been deleted
```

Comment:
> Why are these fallbacks commented out?

my intention was to document that there is still a back up to simply populate the necessary information in case the object is missing. This is true for all places where we are fetching objects i.e locations, groups etc.

> it seems fine to just do this up-front rather than re-fetching the model

The re-fetch is done as a validation to ensure that the message was parsed correctly. Since we are dealing with regex and reading string messages, migrating ~200k records, it would not be possible to do any confident sanity check.

>  Since this is a one-off migration and won't stay in the codebase long

That is correct. I would be deleting these files post run on production.
It was possible to have a separate branch for just the django migration file, merge that and keep these migration files in a separate branch, setup release, run and not merge it at all.
But since they were related i kept them together in the same PR. Happy to go to the other route if that makes sense 
from a better git history.

Reply:
> The re-fetch is done as a validation to ensure that the message was parsed correctly. Since we are dealing with regex and reading string messages, migrating ~200k records, it would not be possible to do any confident sanity check.

That sounds reasonable